### DISCUSSION TITLE
How to implement image diffing for visual regression testing

### DISCUSSION THREAD

**Discussion by rassie** [syrupy-project/syrupy#536] [Q&A]
If I understand the documentation correctly, it's possible to implement a snapshot extension which would compare images visually (e.g. with [pixelmatch](https://pypi.org/project/pixelmatch/)) instead of byte-wise. Is it a matter of writing a custom `SnapshotReporter` and mix it into an extension or would I need to do more?

**Comment by noahnu** [comment]
Same as with the other discussion, sorry about the delayed response.

---

The `SnapshotReporter` is only used for displaying diff results. You'd likely want to override it, specifically the [`diff_lines`](https://github.com/tophat/syrupy/blob/0b4e54488e92c4f28af6fbc49388865cbd946790/src/syrupy/extensions/base.py#L236) method to communicate the results from pixelmatch. The actual assertion unfortunately happens [here](https://github.com/tophat/syrupy/blob/0b4e54488e92c4f28af6fbc49388865cbd946790/src/syrupy/assertion.py#L181) and cannot be overridden at this time.

That being said, it'd be fairly trivial to update this logic to let a custom extension override the equality check. That'd also enable other sorts of "tolerance" assertions. If there is still interest in this, we can move this discussion to a feature request.
